### PR TITLE
Chapter3/homework/5

### PR DESCRIPTION
**Pull request by littleNoodle**

### CODE REVIEW

File: Src/BootCamp.Chapter/ToggleableGridJagged.cs [hunk 1]
```diff
@@ -1,18 +1,52 @@
...
+                        sb.Append("■");
+                    }
+                    else
+                    {
+                        sb.Append(" ");
+                    }
+                }
+                sb.Append(Environment.NewLine);
+            }           
+            var str = sb.ToString().TrimEnd("\r\n".ToCharArray());
```
Comment (Almantask) [inline Src/BootCamp.Chapter/ToggleableGridJagged.cs]: Split to a function of its own

File: Src/BootCamp.Chapter/ToggleableGridJagged.cs [hunk 2]
```diff
@@ -1,18 +1,52 @@
...
         public ToggleableGridJagged(bool[][] toggles, IGridClearer gridClearer)
         {
+            _toggle = toggles;
             _gridClearer = gridClearer;
         }
 
         public void Toggle(int x, int y)
         {
+            _toggle[x][y] = !_toggle[x][y];
+            StringBuilder sb = new StringBuilder();
```
Comment (Almantask) [inline Src/BootCamp.Chapter/ToggleableGridJagged.cs]: Move this close to where it's used- next to loop

File: Src/BootCamp.Chapter/ToggleableGridJagged.cs (hunk 1 above)
Comment (Almantask) [reply Src/BootCamp.Chapter/ToggleableGridJagged.cs]: So the function will encapsulate sb- it's a low level detail

Comment (Almantask) [reply Src/BootCamp.Chapter/ToggleableGridJagged.cs]: Avoid mixing high level functions with implementation of other functions.

**Review by Almantask — CHANGES_REQUESTED** [review]
Try splitting the rendering of a grid from a logical representation and triggering toggles. When you couple both, you have quite a lot of duplication

**Comment by Almantask** [comment]
Please reply to remarks by telling how you resolved them.

**Comment by littleNoodle** [comment]
> Please reply to remarks by telling how you resolved them.

File: Src/BootCamp.Chapter/ToggleableGridJagged.cs (hunk 1 above)
Comment (littleNoodle) [reply Src/BootCamp.Chapter/ToggleableGridJagged.cs]: Sorry for late response.
For this issue, I regard 2D array as special jagged array and define DisplayGrid class to reuse.
Also spend some time to solve the null reference exception about how to make 2D array works.
Through this I could handle 2D array initialization

File: Src/BootCamp.Chapter/ToggleableGridJagged.cs (hunk 2 above)
Comment (littleNoodle) [reply Src/BootCamp.Chapter/ToggleableGridJagged.cs]: I leave the line 19 at Toggle function and make Display function only judge element value( true/false) of array.
High/Low level function is a great strategy, I should focus on code reuse. 
Have a good day~

**Comment by littleNoodle** [comment]
> Please reply to remarks by telling how you resolved them.
I add the reponse above.
Happy new year !

**Review by Almantask — APPROVED** [review]

**Pull request closed without merging**